provisioners/bootstrap_posix.sh

```diff
@@ -0,0 +1,189 @@
+#!/bin/bash
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Purpose: bootstrap a posix host from post install (or image) to a complete puppet run
+# This script it intend to be run either run by hand or from an init system after a host has been
+# provisioned or re-imaged
+
+# Prerequisites:
+#  * curl
+#  * tar w/gzip
+#  * Puppet agent 6.x.x
+#  * R10k 3.x.x
+
+
+# TODO:
+# pull and generate vault secrets to yaml (or json).  Block if secrets don't exist, soft fail if cached
+
+# URL of puppet repo to download
+# TODO: change this url to track master on the moz platform ops org relop
+PUPPET_REPO='https://github.com/dividehex/ronin_puppet/archive/bootstrap.tar.gz'
+
+# If something fails hard, either exit for interactive or hang for non-interactive
+function fail {
+    # TODO: emit an critical error event so provisioner knows this node needs to be handled
+    echo "${@}"
+    # If this is a non-interactive session, hang indefinitely instead of exiting
+    if [ "$NONINTERACTIVE" = true ]; then
+        echo "Hanging..."
+        while true; do sleep 3600; done
+    fi
+    exit 1
+}
+
+# Reset in case getopts has been used previously in the shell.
+OPTIND=1
+
+# Parse options. Use of -l flags the script as non-interactive
+while getopts ":h?l:" opt; do
+    case "$opt" in
+        h|\?)
+            echo "Usage: ./bootstrap_posix.sh -h               - Show help"
+            echo "       ./bootstrap_posix.sh -l /path/logfile - Log output to file"
+            echo "       ./bootstrap_posix.sh                  - Interactive mode"
+            exit 0
+            ;;
+        l)
+            LOG_PATH=$OPTARG
+            NONINTERACTIVE=true
+            # touch log file to see if we have write access and fail otherwise
+            touch "$LOG_PATH" || fail "Can't write log to ${LOG_PATH}"
+            exec >"$LOG_PATH" 2>&1
+            ;;
+        :)
+            echo "Option -$OPTARG requires an argument" >&2
+            exit 1
+            ;;
+    esac
+done
```

Comment:
Nice. This script is clean and clear; Much better than how I'd write it at first pass -- I would not have provided the help or done any pretty opt handling (screw the user)!